core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -539,7 +539,7 @@ class TransactionCoordinator(brokerId: Int,
           s"${txnIdAndPidEpoch.transactionalId} due to timeout")
 
       case error@(Errors.INVALID_PRODUCER_ID_MAPPING |
-                  Errors.INVALID_PRODUCER_EPOCH |
```

Comment:
In the new broker version, when do we still return `INVALID_PRODUCER_EPOCH` then? Or would we never return it any more?

Reply:
We do return `INVALID_PRODUCER_EPOCH` when the client is in old version, as they don't recognize the new error code. See `KafkaApis`

Reply:
My confusion was that for new broker <-> new client communication, could we still return `INVALID_PRODUCER_EPOCH` as the error code? From KIP-588 my understanding is that, txn-coordinator -> producer would not return `INVALID_PRODUCER_EPOCH` anymore for all txn-related requests, and only broker -> producer ProduceResponse would return `INVALID_PRODUCER_EPOCH`. Now I get that you are maintaining the javadoc / handling for compatibility. So we are good :)